File: test/features/pages/data/announcement_presentation_service_test.dart
import 'dart:convert';

import 'package:clashkingapp/core/models/notification_preferences.dart';
import 'package:clashkingapp/core/services/notification_preferences_service.dart';
import 'package:clashkingapp/features/pages/data/announcement_presentation_service.dart';
import 'package:clashkingapp/features/pages/models/app_announcement.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

const testAnnouncement = AppAnnouncement(
  id: 'announcement-1',
  version: '2',
  title: 'Test update',
  subtitle: 'Test story',
  storyUrl: 'https://example.com/story.html',
);

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    SharedPreferences.setMockInitialValues({
      NotificationPreferencesService.localKey: jsonEncode(
        const NotificationPreferences(
          deviceId: 'device-1',
          announcements: true,
        ).toLocalJson(),
      ),
    });
  });

  test(
    'shows an announcement until its ID and version are dismissed',
    () async {
      final service = AnnouncementPresentationService();
      const announcement = testAnnouncement;

      expect(await service.shouldPresent(announcement), isTrue);

      await service.markDismissed(announcement);

      expect(await service.shouldPresent(announcement), isFalse);
    },
  );

  test(
    'a new announcement version is shown after an earlier version',
    () async {
      final service = AnnouncementPresentationService();
      const versionOne = testAnnouncement;
      const versionTwo = AppAnnouncement(
        id: 'announcement-1',
        version: '3',
        title: 'Anime Fury',
        subtitle: 'Updated story',
      );

      await service.markDismissed(versionOne);

      expect(await service.shouldPresent(versionTwo), isTrue);
    },
  );

  test('respects the app announcement notification preference', () async {
    SharedPreferences.setMockInitialValues({
      NotificationPreferencesService.localKey: jsonEncode(
        const NotificationPreferences(
          deviceId: 'device-1',
          events: true,
        ).toLocalJson(),
      ),
    });
    final service = AnnouncementPresentationService();

    expect(await service.shouldPresent(testAnnouncement), isFalse);
  });
}

File: expo/src/features/home/data/announcement-presentation-service.test.ts
import {
  createDefaultNotificationPreferences,
  serializeNotificationPreferencesForLocalStorage,
} from '../../../core/dto/notification-preferences';
import { STORAGE_KEYS } from '../../../core/storage/storage';
import type { StringStore } from '../../../services/storage/auth-storage';
import { AnnouncementPresentationService } from './announcement-presentation-service';
import { AppAnnouncement } from './app-announcement';

class MemoryStore implements StringStore {
  readonly values = new Map<string, string>();
  async getItem(key: string) {
    return this.values.get(key) ?? null;
  }
  async setItem(key: string, value: string) {
    this.values.set(key, value);
  }
  async removeItem(key: string) {
    this.values.delete(key);
  }
}

const announcement = (version: string) =>
  new AppAnnouncement(
    'announcement-1',
    'Anime Fury',
    'June update',
    version,
    null,
    null,
    null,
    'https://cdn.example.com/story.html',
    null,
  );

describe('AnnouncementPresentationService', () => {
  test('respects the announcement preference and versioned dismissal', async () => {
    const store = new MemoryStore();
    const enabled = {
      ...createDefaultNotificationPreferences('device-1', 'production'),
      announcements: true,
    };
    await store.setItem(
      STORAGE_KEYS.notificationSettings,
      JSON.stringify(serializeNotificationPreferencesForLocalStorage(enabled)),
    );
    const service = new AnnouncementPresentationService(store);

    await expect(service.shouldPresent(announcement('2'))).resolves.toBe(true);
    await service.markDismissed(announcement('2'));
    await expect(service.shouldPresent(announcement('2'))).resolves.toBe(false);
    await expect(service.shouldPresent(announcement('3'))).resolves.toBe(true);
  });

  test('defaults to hidden without a persisted opt-in', async () => {
    await expect(
      new AnnouncementPresentationService(new MemoryStore()).shouldPresent(announcement('2')),
    ).resolves.toBe(false);
  });
});
